ambry-commons/src/main/java/com/github/ambry/compression/CompressionFactory.java

```diff
@@ -0,0 +1,130 @@
+/**
+ * Copyright 2022 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.compression;
+
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * CompressionFactory is a compression algorithm registration keyed by the algorithm name.
+ * The factory contains a static shared instance with all supported algorithms pre-registered.
+ *
+ * Factory APIs include:
+ * - Register a new compression algorithm.  The shared instance has pre-registered a set of compression algorithms.
+ * - Get a registered compression by name.
+ * - Get default compressor.
+ * - Set default compressor.
+ * - Get decompressor by compressed buffer generated by compress().
+ */
+public class CompressionFactory {
+
+  // The static instance of compression factory.
+  private static volatile CompressionFactory defaultFactory;
+
+  /**
+   * Get the shared instance of compression factory.
+   * The shared instance contains a pre-registered set of supported compression algorithms.
+   * @return The shared compression factory instance.
+   */
+  public static CompressionFactory getDefault() {
+    if (defaultFactory != null) return defaultFactory;
+
+    synchronized(CompressionFactory.class) {
+      if (defaultFactory == null) {
```

Comment:
Before we start a process (either frontend or server), we know which compression algorithm(s) to use. Therefore, I would recommend to take the compression class names only from config. Ideally there shouldn't be any case where we don't know what `Compression` to use before starting the process.

Reply:
This factory is used for decompression.  It registers all the support decompression.  The compression algorithm is set in setDefaultCompressor().  I do not think the we should select the default compressor in config either.  I discussed my point of view about config above.  Let's chat if we need to make it a config.